cirq/value/random_state.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import cast, Optional, Union
+from typing_extensions import Protocol
+
+import numpy as np
+
+RANDOM_STATE_LIKE = Optional[Union[np.random.RandomState, int]]
+
+
+class PseudoRandNumGen(Protocol):
+
+    def choice(self, *args, **kwargs):
```

Comment:
this is going to be annoying for IDE autocomplete and doc hints. Can't you just make `parse_random_state` return a `RandomState`. I know `np.random` technically is a *module* and not a `RandomState`, but it isn't anything a well placed `cast` can't fix. 

You can also do the other thing of actually extracting the numpy global `RandomState`, but it involves accessing private members

Reply:
I agree what we don't need to make this new class. We can use `types.ModuleType` for the type of `np.random`.

Reply:
well I think one of the advantages of the below-function would be to return an object with a consistent type. I really wish numpy would just expose their global `RandomState` with public methods!

Reply:
Not that this really solves the issue, but we can define a type alias `PRNG_LIKE = Union[np.random.RandomState, types.ModuleType]`

Reply:
I went with casting. While `PRNG_LIKE` would be more "honest" with mypy, the premise is that `np.random` quacks like a `RandomState`.

Reply:
![EFleC2FXkAEphM1](https://user-images.githubusercontent.com/4967059/67443014-f208e400-f5b7-11e9-87b2-92293872fbee.jpg)

Reply:
I don't think we should cast it to a RandomState when it isn't one. This could lead, for instance, to users accidentally changing the global random state by calling `prng.seed` because they think `prng` is a RandomState when really it is `np.random`.

Reply:
I don't think the type annotation is the appropriate place to "document" this caveat. For one thing, both types have a `seed` method that does the same thing so it doesn't disambiguate it at all.

Reply:
I'm not sure what you mean by the `seed` method doing the same thing; in one case it changes the global seed and in another it only changes the seed of the particular RandomState instance. In any case that was just one example of what could go wrong in lying about the type; I can't think of another example for this particular case right now but it still seems like a bad idea to me.

Reply:
In one case it seeds the global `RandomState` and the other seeds a local `RandomState`. The only caveat is that *technically* the global `RandomState` is a module composed entirely of functions that defer to the global `RandomState` object which is hidden with a preceding underscore. 

It's the difference between `x = cirq.CNOT` and `x = cirq.CNotPowGate()`. They're both of type `CNotPowGate`

Reply:
and we're using this helper function to paper over the technicality and assert that global and local random states share a type, which they basically do

Reply:
Maybe @Strilanc can weigh in.

Reply:
Technically it's already the case that the local RandomState instances you get may be used in lots of other places, and calling `seed` on it can cause problems. If someone calls `seed` on an argument they were given then they are probably doing something wrong *anyways*. You don't mutate your arguments without clear naming conventions and documentation saying that you are.

The only reason it's a little iffy here is that we're the ones silently passing the global object instead of a new local instance. The main worry would be that the user infers we are creating independent instances when no seed or random state instance is specified, thinks they can seed it, and then shoots themselves in the foot. E.g. they might have some utility method that seeds random data, that they call all the time, effectively making their computation re-use the same random numbers again and again.

We could use a wrapper object that intercepted the call to seed and raised an error.

We could use a "copy on write" wrapper object that switched to a real RandomState if you called seed.

Probably the simplest thing is to just let them shoot themselves in the foot and then tell them not to use the seed method :l